File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool helper(vector<int>& nums, int index, int target,
                vector<vector<int>>& dp) {
        if (target < 0)
            return false;
        if (target == 0)
            return true;
        if (index == 0) {
            if (target == 0 || nums[index] == target)
                return true;
            return false;
        }
        if (dp[index][target] != -1)
            return dp[index][target];
        //
        target -= nums[index];
        bool left = helper(nums, index - 1, target, dp);
        target += nums[index];
        bool right = helper(nums, index - 1, target, dp);
        return dp[index][target] = left || right;
    }

    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (sum & 1)
            return false;
        sum = sum / 2;
        int n = nums.size();

        vector<vector<int>> dp(n + 1, vector<int>(sum + 1, -1));
        return helper(nums, n - 1, sum, dp);
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
      public boolean canPartition(int[] nums) {
        int sum = 0;
        int n = nums.length;
        for (int i : nums)
            sum += i;
        if (sum % 2 != 0)
            return false;
        int target = sum / 2;
        boolean[][] dp = new boolean[n + 1][target + 1];
        for (int i = 0; i <= n; i++) {
            dp[i][0] = true;
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= target; j++) {
                boolean pick = false;
                if (nums[i-1] <= j) {
                    pick = dp[i - 1][j - nums[i-1]];
                }
                boolean notPick = dp[i - 1][j];
                dp[i][j] = pick || notPick;
            }
        }
        return dp[n][target];
        // return helper(nums, n - 1, target, dp);
    }
}